source/simulator/stokes_matrix_free.cc

```diff
@@ -1696,6 +1880,100 @@ namespace aspect
+                                   sim.mpi_communicator);
+
+            QGauss<dim>  quadrature_formula(sim.parameters.stokes_velocity_degree+1);
+            FEValues<dim> fe_values (fe_v, quadrature_formula,
+                                     update_values   | update_gradients |
+                                     update_quadrature_points | update_JxW_values);
+
+            const unsigned int   dofs_per_cell   = fe_v.dofs_per_cell;
+            const unsigned int   n_q_points      = quadrature_formula.size();
+
+            FullMatrix<double>   cell_matrix (dofs_per_cell, dofs_per_cell);
+
+            std::vector<types::global_dof_index> local_dof_indices (dofs_per_cell);
+            const FEValuesExtractors::Vector velocities (0);
+
+            std::vector<SymmetricTensor<2,dim> > symgrad_phi_u (dofs_per_cell);
+
+            ConstraintMatrix boundary_constraints;
+            boundary_constraints.reinit(locally_relevant_dofs);
+            boundary_constraints.add_lines (mg_constrained_dofs.get_refinement_edge_indices(level));
+            boundary_constraints.add_lines (mg_constrained_dofs.get_boundary_indices(level));
+#if DEAL_II_VERSION_GTE(9,2,0)
+            boundary_constraints.merge(mg_constrained_dofs.get_user_constraint_matrix(level));
+#endif
+            boundary_constraints.close();
+
+            typename DoFHandler<dim>::level_cell_iterator cell = dof_handler_v.begin(level),
+                                                          endc = dof_handler_v.end(level);
+            for (; cell!=endc; ++cell)
+              if (cell->level_subdomain_id()==sim.triangulation.locally_owned_subdomain())
+                {
+                  cell_matrix = 0;
+                  fe_values.reinit (cell);
+
+
+                  typename DoFHandler<dim>::level_cell_iterator DG_cell(&(sim.triangulation),
+                                                                        level,
+                                                                        cell->index(),
+                                                                        &dof_handler_projection);
+                  std::vector<types::global_dof_index> dg_dof_indices(dof_handler_projection.get_fe(0).dofs_per_cell);
+                  DG_cell->get_active_or_mg_dof_indices(dg_dof_indices);
+                  double viscosity = level_coef_dof_vec[level](dg_dof_indices[0]);
+
+                  for (unsigned int q=0; q<n_q_points; ++q)
+                    {
+                      for (unsigned int k=0; k<dofs_per_cell; ++k)
+                        symgrad_phi_u[k] = fe_values[velocities].symmetric_gradient (k, q);
+
+                      for (unsigned int i=0; i<dofs_per_cell; ++i)
+                        for (unsigned int j=0; j<dofs_per_cell; ++j)
+                          cell_matrix(i,j) += (2*viscosity*(symgrad_phi_u[i]*symgrad_phi_u[j])
+                                               * fe_values.JxW(q));
```

Comment:
Not sure how expensive this whole operation is, but in the assembly we typically pull `JxW`out of that double loop.

Reply:
good point!